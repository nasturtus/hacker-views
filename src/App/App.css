@import url("https://fonts.googleapis.com/css?family=Poppins:100,900");

body {
  margin: 0 auto;
  max-width: 1000px;
  color: black;
  background: white;
}

.container-grid {
  display: grid;
  /* margin-left: 5%;
  margin-right: 5%; */
  grid-row-gap: 20px;
}

.interactions {
  text-align: center;
}

.header {
  text-align: right;
  font-family: "Poppins", sans-serif;
}

.title {
  font-weight: 100;
  font-size: 300%;
  margin-bottom: 0%;
}

.subtitle {
  font-weight: 900;
  font-weight: 200%;
  margin-top: 0%;
  color: rgb(0, 0, 255);
}


ul,
li {
  list-style: none;
  padding: 0px;
  margin: 2px;
}

input {
  padding: 10px;
  border-radius: 5px;
  outline: none;
  margin-right: 10px;
  border: 1px solid #dddddd;
}

button {
  padding: 5px;
  border-radius: 5px;
  border: 1px solid #dddddd;
  background: transparent;
  color: #808080;
  cursor: pointer;
}

button:hover {
  color: #222;
}

*:focus {
  outline: none;
}

.loader {
  box-sizing: border-box;
  width: 50px;
  height: 50px;
  position: absolute;
  top: 50%;
  left: 45%;
  border-radius: 50%;
  border: solid 5px gray;
  border-top-color: yellow;
  animation-name: spinner;
  animation-duration: 1s;
  animation-timing-function: linear;
  animation-iteration-count: infinite;
  /* alternatively: animation: spinner 1.0s linear infinite; */
}

@keyframes spinner {
  to {
    transform: rotate(360deg);
  }
}
