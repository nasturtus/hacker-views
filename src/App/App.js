// external libraries
import React, { Component } from "react";
import axios from "axios";

// app imports
import { Search } from "../Search";
import { Table } from "../Table";
import { searchUrl, frontPageUrl } from "../Utils";

// css
import "./App.css";

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      result: {
        hits: [],
        page: 0,
        nbPages: 0,
        isLoading: false
      },
      searchTerm: "",
      error: null
    };
    this.setSearchTopStories = this.setSearchTopStories.bind(this);
    this.searchChange = this.searchChange.bind(this);
    this.onSearchSubmit = this.onSearchSubmit.bind(this);
    this.fetchSearchTopStories = this.fetchSearchTopStories.bind(this);
  }

  componentDidMount() {
    const { searchTerm } = this.state;
    const page = this.state.result.page;
    this.fetchSearchTopStories(searchTerm, page);
  }

  render() {
    console.log("in render..");
    console.log(this.state.result.hits);
    const { result, searchTerm } = this.state; // object destructuring

    const tableDOM = (
      <Table
        hits={result.hits}
        page={result.page}
        nbPages={result.nbPages}
        searchTerm={searchTerm}
        filterCallback={this.isSearched}
        loadMoreStories={this.fetchSearchTopStories}
      />
    );

    if (this.state.result.isLoading) {
      return <div className="loader" />;
    }
    return (
      <div className="container-grid">
        <div className="header">
          <h1 className="title"> Hacker Views</h1>
          <h3 className="subtitle">hacker news plus great views.</h3>
        </div>
        <div className="interactions">
          <Search
            value={searchTerm}
            onChange={this.searchChange}
            onSubmit={this.onSearchSubmit}
          />
        </div>
        {/* option 1 for conditional rendering: using ternary operator */}
        {/* {result ? tableDOM : noResultsDOM} */}
        {/* option 2 for conditional rendering: using && operator. more concise and obviates need for noResultsDOM */}
        {result && tableDOM}
      </div>
    );
  }

  // remember setState does not immediately execute.
  // hence async-await to ensure fetchSearch... is called only after state is set
  async onSearchSubmit(event) {
    event.preventDefault();

    // restart search from page 0 and reinitialise hits array to empty before new search
    await this.setState({ result: { hits: [], page: 0 } });

    this.fetchSearchTopStories(this.state.searchTerm, this.state.result.page);
  }

  inCache() {
    return this.state.cache[this.state.searchTerm];
  }

  fetchSearchTopStories(searchTerm, page) {
    //!!important: when setting state to an object that has multiple key-value pairs
    // you have to set state to each key-value pair.
    // setting the state of only one will leave the others undefined.
    // this.setState({
    //   result: {
    //     hits: this.state.result.hits,
    //     page: this.state.result.page,
    //     nbPages: this.state.result.nbPages,
    //     isLoading: true
    //   }
    // });
    // line 134-35 using destructuring is  a shortened way of achieving the same as above
    console.log(`Fetch page number ${page} for search term ${searchTerm}`);
    let url = null;
    const pageNumber = "&page=" + page.toString();

    if (searchTerm === "") {
      // set url to fetch front_page stories
      url = frontPageUrl + pageNumber;
    } else {
      // set url to fetch pages related to search term submitted
      searchTerm += pageNumber;
      url = searchUrl + searchTerm;
      console.log({ url });
    }
    const updatedResult = { ...this.state.result, isLoading: true };
    this.setState({ result: updatedResult });

    axios(url)
      .then(results => {
        console.log(results);
        this.setSearchTopStories(
          results.data.hits,
          results.data.page,
          results.data.nbPages
        );
      })
      .catch(error => this.setState({ error }));
  }

  setSearchTopStories(hits, page, nbPages) {
    // the following line is only if you want to add new results to existing results.
    const updatedHits = [...this.state.result.hits, ...hits];

    this.setState({
      result: {
        hits: updatedHits,
        page: page,
        nbPages: nbPages,
        isLoading: false
      }
    });
  }

  searchChange(e) {
    this.setState({ searchTerm: e.target.value.toLowerCase() });
  }

  isSearched(searchTerm) {
    // item is the argument that filter receives, which is automatically passed to a filter's callback
    return item => item.title.toLowerCase().includes(searchTerm);
  }
}

export { App, Search, Table };
