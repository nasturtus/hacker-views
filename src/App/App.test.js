import React from "react";
import ReactDOM from "react-dom";
import { App, Search, Table } from "./App";
import renderer from "react-test-renderer";

describe("App", () => {
  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });
  it("has a valid snapshot", () => {
    const component = renderer.create(<App />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

// describe("Search component", () => {
//   const props = {
//     value: "dd",
//     oChange: () => {},
//     onSubmit: () => {}
//   };
//   it("renders without crashing", () => {
//     const div = document.createElement("div");
//     ReactDOM.render(<Search {...props}>Search</Search>, div);
//     ReactDOM.unmountComponentAtNode(div);
//   });
//   it("has a valid snapshot", () => {
//     const component = renderer.create(<Search>Search</Search>);
//     let tree = component.toJSON();
//     expect(tree).toMatchSnapshot();
//   });
// });

// describe("Test component", () => {
//   const props = {
//     list1: [
//       {
//         title: "Some title",
//         pages: 100,
//         author: "Some author"
//       }
//     ],
//     list2: [
//       {
//         title: "Some title",
//         pages: 100,
//         author: "Some author"
//       }
//     ]
//   };
//   it("renders without crashing", () => {
//     const div = document.createElement("div");
//     ReactDOM.render(<Table {...props} />, div);
//     ReactDOM.unmountComponentAtNode(div);
//   });
// });
