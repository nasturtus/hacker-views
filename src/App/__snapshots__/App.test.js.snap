// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App has a valid snapshot 1`] = `
<div
  className="page"
>
  <div
    className="interactions"
  >
    <form
      onSubmit={[Function]}
    >
      <h1>
        Search
      </h1>
      <input
        onChange={[Function]}
        type="text"
        value="purescript"
      />
    </form>
  </div>
  <div
    className="table"
  >
    <h1>
      Results
    </h1>
  </div>
</div>
`;

exports[`Search component has a valid snapshot 1`] = `
<form
  onSubmit={undefined}
>
  <h1>
    Search
  </h1>
  <input
    onChange={undefined}
    type="text"
    value={undefined}
  />
</form>
`;
